File: 2022/2022W20/preppindataW20.R
library(readxl)
library(dplyr)
library(stringi)
library(stringr)

xlsx <- "C:/Data/PreppinData/TC22 Input.xlsx"

final <-  read_excel(xlsx, 'Registrations') %>%
  group_by(`Email`) %>%
  mutate('Online/In Person' = stri_replace_all_regex(`Online/In Person`, pattern = c('^O.*', '^I.*'), 
                                                     replacement = c('Online', 'In Person'), vectorize_all = FALSE),
         'Company' = str_extract_all(`Email`, '(?<=@)([^\\.]+)'),
         'Planning to attend' = n_distinct(`Session ID`)) %>%
  merge(read_excel(xlsx, 'Sessions'), by='Session ID') %>%
  anti_join(read_excel(xlsx, 'In Person Attendees'), by = c('Session', 'First Name', 'Last Name')) %>%
  anti_join(read_excel(xlsx, 'Online Attendees'), by = c('Session', 'Email')) %>%
  group_by(`Email`) %>%
  mutate('Not Attended %' = n_distinct(`Session ID`)/`Planning to attend`*100) %>%
  rename('Session not attended' = 'Session') %>%
  select(c('Company', 'First Name', 'Last Name', 'Email', 'Online/In Person', 'Session not attended', 'Not Attended %'))
  
View(final)

File: 2022/2022W20/PreppindataW20.py
import pandas as pd
import re

xlsx = pd.ExcelFile(r"C:\Data\PreppinData\TC22 Input.xlsx")

regs = pd.read_excel(xlsx, 'Registrations')
regs['Online/In Person'] = regs['Online/In Person'].str.replace('^O.*','Online',regex=True)
regs['Online/In Person'] = regs['Online/In Person'].str.replace('^I.*','In Person',regex=True)

regs['Company'] = regs['Email'].apply(lambda x: re.search('(?<=@)([^\.]+)', x).group(1))
regs['Planning to attend'] = regs['Session ID'].groupby(regs['Email']).transform('nunique') 
regs = regs.merge(pd.read_excel(xlsx, 'Sessions'), on='Session ID')

not_att = regs.merge(pd.read_excel(xlsx, 'In Person Attendees'), 
                     on=['Session', 'First Name', 'Last Name'], 
                     how = 'outer', indicator = True).query('_merge == "left_only"').drop('_merge', axis=1)
not_att = not_att.merge(pd.read_excel(xlsx, 'Online Attendees'), 
                        on=['Session', 'Email'], 
                        how = 'outer', indicator = True).query('_merge == "left_only"')

not_att['Not attended'] = not_att['Session ID'].groupby(regs['Email']).transform('nunique') 
not_att['Not Attended %'] = not_att['Not attended']/not_att['Planning to attend']*100
not_att = not_att.rename(columns={'Session':'Session not attended'})

not_att = not_att[['Company', 'First Name', 'Last Name', 'Email', 'Online/In Person', 'Session not attended', 'Not Attended %']]
